File: python/dsa/data-structures/binary-tree/symmetric-tree.py
# Definition for a binary tree node.
# class TreeNode: val, left, right

def isSymmetric(self, root): # @viz: is_symmetric_start
    if not root: # @viz: is_symmetric_base
        return True # @viz: is_symmetric_base
    return self.isMirror(root.left, root.right) # @viz: is_symmetric_call

def isMirror(self, t1, t2): # @viz: is_mirror_start
    if not t1 and not t2: # @viz: base_both_null
        return True # @viz: base_both_null
    if not t1 or not t2 or t1.val != t2.val: # @viz: base_mismatch
        return False # @viz: base_mismatch

    return self.isMirror(t1.left, t2.right) and \
           self.isMirror(t1.right, t2.left) # @viz: recurse_inner

File: java/dsa/data-structures/binary-tree/symmetric-tree.java
/** Definition for a binary tree node */
// class TreeNode { int val; TreeNode left; TreeNode right; }

public boolean isSymmetric(TreeNode root) { // @viz: is_symmetric_start
    if (root == null) { // @viz: is_symmetric_base
        return true; // @viz: is_symmetric_base
    }
    return isMirror(root.left, root.right); // @viz: is_symmetric_call
}

private boolean isMirror(TreeNode t1, TreeNode t2) { // @viz: is_mirror_start
    if (t1 == null && t2 == null) { // @viz: base_both_null
        return true; // @viz: base_both_null
    }
    if (t1 == null || t2 == null || t1.val != t2.val) { // @viz: base_mismatch
        return false; // @viz: base_mismatch
    }

    return isMirror(t1.left, t2.right) && // @viz: recurse_outer
           isMirror(t1.right, t2.left); // @viz: recurse_inner
}